### ISSUE TITLE
how to make a prediction in C++

### ISSUE THREAD

**Issue by onauparc** [BVLC/caffe#499]
Hi,
I'm using caffe and I'm trying to make a prediction in C++ with a pretrained network like this python example  http://nbviewer.ipython.org/github/BVLC/caffe/blob/master/examples/imagenet_classification.ipynb. But I don't find a C++ method for prediction and I don't get how to pass an input image and make the prediction. I've seen test_net.cpp but it does not pass any image in input. How can I pass an image or more images in input to the network? How can I make a prediction?

**Comment by joeupwu** [comment]
Try to use MemoryData layer as input to prevent fake image list.

fragment of input net:
layer {
  name: "input"
  type: "MemoryData"
  top: "data"
  top: "label"

  transform_param {
    mirror: false
    crop_size: 227
    mean_file: "d:/project/caffe/data/ilsvrc12/imagenet_mean.binaryproto"  
  }
  
  memory_data_param {
  	batch_size: 1
  	channels: 3
  	height: 227
  	width: 227
  }  
}

fragment of C++ code:
	float loss;
	Net<float> caffe_net(NETMODELFILE, caffe::TEST);
	caffe_net.CopyTrainedLayersFrom(TRAINEDFILE);

	// load image files, resize and added to vector
	cv::Mat image = cv::imread(INIMAGEFILE);
	cv::Mat rszimage;

	// The mean file image size is 256x256, need to resize the input image to 256x256
	cv::resize(image, rszimage, cv::Size(256, 256));

	std::vector<cv::Mat> images(1, rszimage);
	std::vector<int> labels(1, 0);

	shared_ptr<MemoryDataLayer<float>> md_layer =
		boost::dynamic_pointer_cast <MemoryDataLayer<float>>(caffe_net.layers()[0]);
	if (!md_layer) {
		std::cout << "The first layer is not a MemoryDataLayer!\n";
		return -1;
	}

	md_layer->AddMatVector(images, labels);

	caffe_net.ForwardPrefilled(&loss);
	std::cout << "loss: " << loss << "\n";
        // read the 'prob' layer and get the result
	shared_ptr<Blob<float>> prob = caffe_net.blob_by_name("prob");
	float maxval= 0;
	int   maxinx= 0;
	for (int i = 0; i < prob->count(); i++)
	{
		float val= prob->cpu_data()[i];
		if (val> maxval)
		{
			maxval= val;
			maxinx= i;
		}
		std::cout << "[" << i << "]" << val<< "\n";
	}
	std::cout << "Max value = " << maxval<< ", Max index = " << maxinx<< "\n";

**Comment by anukuppan** [comment]
How to train the MINST test data .Can any one suggest any tutorials for training the data sets ?
Thanks in advance.




Regards,
Anu kuppan

**Comment by antran89** [comment]
What do you mean training on test data?

On Monday, October 26, 2015, anukuppan <notifications@github.com> wrote:

> How to train the MINST test data .Can any one suggest any tutorials for
> training the data sets ?
> Thanks in advance.
>
> Regards,
> Anu kuppan
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/issues/499#issuecomment-151114018>.
>


-- 
---------------------------------------------------------------------------
Full name: Tran Lam An
NUS graduate student
Email: tranlaman at gmail dot com
Homepage: http://antran89.github.io/
---------------------------------------------------------------------------

**Comment by anukuppan** [comment]
How to test  caffe classifier trained by MNIST data sets?

**Comment by cmanders** [comment]
Hello, 
if my input is:
name: "ZF"
input: "data"
input: "im_info"
state {
  phase: TEST
}
input_shape {
  dim: 1
  dim: 3
  dim: 224
  dim: 224
}
input_shape {
  dim: 1
  dim: 3
}

how do I push-back the needed blobs before calling forward?

Thanks!
Corey

**Comment by Arieal** [comment]
How to implement caffe into my C++ application an example?How to run classfication.cpp ?

**Comment by antran89** [comment]
@Arieal You can refer to the example documents http://caffe.berkeleyvision.org/gathered/examples/cpp_classification.html.

**Comment by Arieal** [comment]
@antran89  Thanks very much! what I meant is how to link caffe as library and call functions in my own project not modify the  example already configured and build successfully in caffe.

**Comment by antran89** [comment]
@Arieal I think it is not difficult, link again libcaffe.so and include some .h files should be enough.
